File: typepos.ts
interface OrderLine {
  unitPrice: number;
  quantity: number;
}

function totalPrice(lines: Iterable<OrderLine>): number {
  let total = 0;
  for (let line of lines) {
    total += line.unitPrice * line.quantity;
  }
  return total;
  // return lines.reduce(
  //   (total, line) => total + line.unitPrice * line.quantity,
  //   0,
  // );
}

function main() {
  let line: unknown = { unitPrice: 0.25, quantity: 2 };
  let order: OrderLine[] = [
    // <OrderLine> line,
    line as OrderLine,
    { unitPrice: 1.00, quantity: 1 },
  ];
  console.log(totalPrice(order));
}

main()

File: TypePos.java
import java.math.*;
import java.util.*;
import java.util.stream.*;

class OrderLine {
    BigDecimal unitPrice;
    int quantity;
    OrderLine(BigDecimal unitPrice, int quantity) {
        this.unitPrice = unitPrice;
        this.quantity = quantity;
    }
}

class TypePos {
    static BigDecimal totalPrice(Iterable<OrderLine> lines) {
        return StreamSupport.stream(lines.spliterator(), false)
            .map(line -> line.unitPrice.multiply(new BigDecimal(line.quantity)))
            .reduce(BigDecimal.ZERO, (a, b) -> a.add(b));
    }

    public static void main(String[] args) {
        Object line = new OrderLine(new BigDecimal("0.25"), 2);
        var order = Arrays.asList(
            (OrderLine)line,
            new OrderLine(new BigDecimal("1.00"), 1)
        );
        System.out.println(totalPrice(order));
    }
}
